pkg/kv/kvserver/cordon_integration_test.go

```diff
@@ -0,0 +1,94 @@
+package kvserver_test
+
+import (
+	"bytes"
+	"context"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverbase"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCordonIfPanicDuringApply(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	skip.UnderShort(t)
+
+	ctx := context.Background()
+
+	dir, cleanupFn := testutils.TempDir(t)
+	defer cleanupFn()
+
+	testArgs := func(dontPanicOnApplyPanicOrFatalError bool) base.TestServerArgs {
+		return base.TestServerArgs{
+			Settings: cluster.MakeClusterSettings(),
+			// We will start up a second server after stopping the first one to
+			// simulate a server restart, since cordoning in response to a panic
+			// during apply only takes effect after a server restart. So we use
+			// the same store for the two servers (they are the same node). This
+			// way, the should cordon write to storage done by the first server is
+			// seen by the second server at startup.
+			StoreSpecs: []base.StoreSpec{{Path: dir + "/store-1"}},
+			Knobs: base.TestingKnobs{
+				Store: &kvserver.StoreTestingKnobs{
+					DontPanicOnApplyPanicOrFatalError: dontPanicOnApplyPanicOrFatalError,
```

Comment:
The need for this awkward testing knob would disappear once we trigger the panic using a special request as suggested above.